components/tikv_util/src/config.rs

```diff
@@ -1456,68 +1433,115 @@ impl RaftDataStateMachine {
-            Self::remove_dir_safe(&self.source);
-            Self::rename_dir_safe(&self.target_staging, &self.target);
-            return None;
-        }
-        if Self::data_exists(&self.source) {
-            // Init -> Migrating
-            assert!(!Self::data_exists(&self.source_staging));
-            Self::rename_dir_safe(&self.source, &self.source_staging);
-        } else if !Self::data_exists(&self.source_staging) {
-            // No source data.
-            return None;
+        if !Self::data_exists(&self.source) {
+            // Recover from Completed state.
+            if self.in_progress_marker.exists() {
+                Self::must_remove(&self.in_progress_marker);
+            }
+            return false;
+        } else if self.in_progress_marker.exists() {
+            if let Some(real_source) = self.read_marker() {
+                // Recover from Migrating state.
+                if real_source == self.target {
+                    assert!(Self::data_exists(&self.target));
+                    Self::must_remove(&self.source);
+                    return false;
+                } else {
+                    assert!(real_source == self.source);
+                    Self::must_remove(&self.target);
+                    return true;
+                }
+            } else {
+                // Halfway between Init and Migrating.
+                assert!(!Self::data_exists(&self.target));
+            }
         }
-        Some(self.source_staging.clone())
+        // Init -> Migrating.
+        self.write_marker();
+        true
     }
 
     /// Exits the `Migrating` state and enters the `Completed` state.
     pub fn after_dump_data(&mut self) {
+        assert!(Self::data_exists(&self.source));
         assert!(Self::data_exists(&self.target));
-        assert!(!Self::data_exists(&self.source));
-        assert!(!Self::data_exists(&self.target_staging));
-        Self::remove_dir_safe(&self.source_staging);
+        Self::must_remove(&self.source); // Enters the `Completed` state.
+        Self::must_remove(&self.in_progress_marker);
     }
 
-    fn data_exists(path: &Path) -> bool {
-        if !path.exists() || !path.is_dir() {
-            return false;
+    // `after_dump_data` involves two atomic operations, insert a check point between
+    // them to test crash safety.
+    #[cfg(test)]
+    fn after_dump_data_with_check<F: Fn()>(&mut self, check: &F) {
+        assert!(Self::data_exists(&self.source));
+        assert!(Self::data_exists(&self.target));
+        Self::must_remove(&self.source); // Enters the `Completed` state.
+        check();
+        Self::must_remove(&self.in_progress_marker);
+    }
+
+    fn write_marker(&self) {
+        use std::io::Write;
+        let mut f = fs::File::create(&self.in_progress_marker).unwrap();
+        f.write_all(self.source.to_str().unwrap().as_bytes())
+            .unwrap();
+        f.sync_all().unwrap();
+        f.write_all(b"//").unwrap();
```

Comment:
Perhaps `##` is a safer mark as `//` is possible to appear in normal path.

Reply:
It appears "#" is allowed in file name, whilst "/" is not, so it's unlikely to have two consecutive slashes in a path.